File: security/spring-boot-social-login/src/main/java/com/sivalabs/bookmarker/domain/service/UserService.java
package com.sivalabs.bookmarker.domain.service;

import com.sivalabs.bookmarker.domain.entity.Role;
import com.sivalabs.bookmarker.domain.entity.User;
import com.sivalabs.bookmarker.domain.model.UserDTO;
import com.sivalabs.bookmarker.domain.repository.RoleRepository;
import com.sivalabs.bookmarker.domain.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Collections;
import java.util.Optional;

import static com.sivalabs.bookmarker.domain.utils.Constants.ROLE_USER;

@Service
@Transactional
@Slf4j
@RequiredArgsConstructor
public class UserService {
    private final UserRepository userRepository;
    private final RoleRepository roleRepository;
    private final PasswordEncoder passwordEncoder;

    @Transactional(readOnly = true)
    public Optional<UserDTO> getUserById(Long id) {
        return userRepository.findById(id).map(UserDTO::fromEntity);
    }

    @Transactional(readOnly = true)
    public Optional<User> getUserByEmail(String email) {
        return userRepository.findByEmail(email);
    }

    @Transactional(readOnly = true)
    public boolean isUserExistsByEmail(String email) {
        return userRepository.existsByEmail(email);
    }

    public UserDTO createUser(UserDTO user) {
        if (userRepository.existsByEmail(user.getEmail())) {
            throw new RuntimeException("Email "+user.getEmail()+" is already in use");
        }
        user.setPassword(passwordEncoder.encode(user.getPassword()));
        User userEntity = user.toEntity();
        Optional<Role> roleUser = roleRepository.findByName(ROLE_USER);
        userEntity.setRoles(Collections.singleton(roleUser.orElse(null)));
        return UserDTO.fromEntity(userRepository.save(userEntity));
    }
}

File: spring-boot-k8s-demo/src/main/kotlin/com/sivalabs/geeksclub/users/UserService.kt
package com.sivalabs.geeksclub.users

import com.sivalabs.geeksclub.entities.User
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

import java.util.Optional

@Service
@Transactional
class UserService(val userRepository: UserRepository, val passwordEncoder: PasswordEncoder) {

    fun findUserByEmail(email: String): Optional<User> {
        return userRepository.findByEmail(email)
    }

    fun createUser(user: User): User {
        user.password=passwordEncoder.encode(user.password)
        user.role = ("ROLE_USER")
        return userRepository.save(user)
    }

    fun findUserById(userId: Long): Optional<User> {
        return userRepository.findById(userId)
    }
}